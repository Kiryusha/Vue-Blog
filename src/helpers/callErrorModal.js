// calling error modal
export default function (context, error) {
  let errText = 'Неизвестная ошибка.';

  if (error.response) {
    errText = error.response.data.message ? error.response.data.message :
      error.response.data;
  } else {
    errText = error.message;
  }

  context.$Progress.fail();
  context.$modal.show('response', { message: errText });
}
