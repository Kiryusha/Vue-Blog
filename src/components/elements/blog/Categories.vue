<template lang="pug">
  Sidebar
    span(slot="title") Категории
    .categories
      .container
        .item(v-for="item in categories")
          Button(
            :class="{'_active':activeCategory === item}"
            :type="'a'"
            :view="'categories'"
            @click="selectCategory(item)"
          ) {{ item }}
</template>

<script>
import { mapState, mapActions } from 'vuex';
import Button from 'Components/elements/general/Button';
import Sidebar from 'Components/elements/blog/Sidebar';
import callErrorModal from '@/helpers/callErrorModal';

export default {
  components: {
    Button,
    Sidebar,
  },
  computed: {
    ...mapState({
      categories: state => state.list.categories,
      activeCategory: state => state.list.activeCategory,
    }),
  },
  methods: {
    ...mapActions([
      'fetchList',
    ]),
    async selectCategory(category) {
      try {
        if (category !== this.activeCategory) {
          await this.fetchList({
            activeCategory: category,
          });
        } else {
          await this.fetchList({});
        }

        if (this.$route.path !== '/blog/') {
          this.$router.push('/blog/');
        }
      } catch (error) {
        callErrorModal(this, error);
      }
    },
  },
};
</script>

<style lang="stylus" scoped>
  .container
    display flex
    flex-wrap wrap
    margin 0 -5px -5px 0

  .item
    margin 0 5px 5px 0
    min-width 47px

</style>
