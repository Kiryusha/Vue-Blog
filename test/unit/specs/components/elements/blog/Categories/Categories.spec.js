// Categories.vue

import { mount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Categories from 'Components/elements/blog/Categories';
import callErrorModal from '@/helpers/callErrorModal';

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock('@/helpers/callErrorModal', () => jest.fn());

describe('Categories.vue', () => {
  const fetchList = jest.fn();
  const config = {
    localVue,
    mocks: {
      $route: {
        path: '/blog/',
      },
      $router: {
        push: jest.fn(),
      },
    }
  };

  let storeConfig = {
    state: {
      list: {
        categories: ['cat','dog'],
        activeCategory: 'cat',
      },
    },
    actions: {
      fetchList: () => {},
    },
  };
  let store = new Vuex.Store(storeConfig);

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('component shallow rendering', async () => {
    it('should match a snapshot', () => {
      const wrapper = mount(Categories, { ...config, store });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('custom actions', () => {
    it('should call plain fetchList() on selectCategory() by selecting active category', () => {
      const wrapper = mount(Categories, { ...config, store, methods: {
        fetchList,
      }});

      wrapper.vm.selectCategory('cat')

      expect(fetchList).toHaveBeenCalledTimes(1);
      expect(fetchList).toHaveBeenCalledWith({});
    });

    it('should call fetchList("dog") on selectCategory() by selecting "dog" category while "cat" is active', () => {
      const wrapper = mount(Categories, { ...config, store, methods: {
        fetchList,
      }});

      wrapper.vm.selectCategory('dog')

      expect(fetchList).toHaveBeenCalledTimes(1);
      expect(fetchList).toHaveBeenCalledWith({
        activeCategory: 'dog',
      });
    });

    it('should call $router.push("/blog/") on selectCategory() if route differs from "/blog/"', async () => {
      config.mocks.$route.path = '/blog/test/';
      const wrapper = mount(Categories, { ...config, store, methods: {
        fetchList,
      }});

      await wrapper.vm.selectCategory('cat')

      expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1);
      expect(wrapper.vm.$router.push).toHaveBeenCalledWith('/blog/');
    });

    it('should call callErrorModal() on selectCategory() if error happens', async () => {
      const fetchList = jest.fn().mockRejectedValue(new Error('test'));
      const wrapper = mount(Categories, { ...config, store, methods: {
        fetchList,
      }});

      await wrapper.vm.selectCategory('cat')

      expect(callErrorModal).toHaveBeenCalledTimes(1);
    });
  });
});
