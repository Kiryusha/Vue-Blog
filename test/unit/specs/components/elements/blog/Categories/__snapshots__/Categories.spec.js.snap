// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories.vue component shallow rendering should match a snapshot 1`] = `
<div class="sidebar">
  <div class="title"><span>Категории</span></div>
  <div class="content">
    <div class="categories">
      <div class="container">
        <div class="item">
          <a class="button _categories _active">
            <div class="caption">cat</div>
            <!---->
            <!---->
            <!---->
            <!---->
            <!---->
            <!---->
            <!---->
          </a>
        </div>
        <div class="item">
          <a class="button _categories">
            <div class="caption">dog</div>
            <!---->
            <!---->
            <!---->
            <!---->
            <!---->
            <!---->
            <!---->
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
